plugins/plugin-ssh-machine/src/main/java/org/eclipse/che/plugin/machine/ssh/SshMachineInstanceProvider.java

```diff
@@ -62,41 +67,58 @@ public String getType() {
         return supportedRecipeTypes;
     }
 
+    /**
+     * Creates instance from scratch or by reusing a previously one by using specified {@link MachineSource}
+     * data in {@link MachineConfig}.
+     *
+     * @param machine
+     *         machine description
+     * @param lineConsumer
+     *         output for instance creation logs
+     * @return newly created {@link Instance}
+     * @throws UnsupportedRecipeException
+     *         if specified {@code recipe} is not supported
+     * @throws InvalidRecipeException
+     *         if {@code recipe} is invalid
+     * @throws NotFoundException
+     *         if instance described by {@link MachineSource} doesn't exists
+     * @throws MachineException
+     *         if other error occurs
+     */
     @Override
-    public Instance createInstance(Recipe recipe,
-                                   Machine machine,
-                                   LineConsumer machineLogsConsumer) throws MachineException {
+    public Instance createInstance(Machine machine, LineConsumer lineConsumer)
+            throws UnsupportedRecipeException, InvalidRecipeException, NotFoundException, MachineException {
         requireNonNull(machine, "Non null machine required");
-        requireNonNull(machineLogsConsumer, "Non null logs consumer required");
+        requireNonNull(lineConsumer, "Non null logs consumer required");
+        requireNonNull(machine.getConfig().getSource().getLocation(), "Location in machine source is required");
```

Comment:
I saw that you commented fixed, so I misunderstood you. I mean that now it is possible to provide either location or content of ssh-machine recipe so we can check that exact one of the fields (location or content) is not null.